File: agent/app/zones/geometry.py
"""Zone geometry on NORMALISED coordinates (Phase 2 §11–12).

Polygons are stored as fractions of the frame (0–1), so a zone drawn on a
640×360 substream still fits the 1920×1080 main stream. Everything here is
pure Python — it runs per track per frame, so no numpy allocations.
"""

from __future__ import annotations

from dataclasses import dataclass

Point = tuple[float, float]

MAX_POINTS = 32
MIN_AREA = 0.0005  # 0.05 % of the frame — smaller is almost certainly a mis-click


class PolygonError(ValueError):
    pass


def validate_polygon(raw: object) -> list[Point]:
    """Input validation for zone polygons (Phase 2 §62). Raises PolygonError."""
    if not isinstance(raw, list) or not 3 <= len(raw) <= MAX_POINTS:
        raise PolygonError(f"A zone needs between 3 and {MAX_POINTS} points.")
    pts: list[Point] = []
    for p in raw:
        try:
            x, y = (float(p["x"]), float(p["y"])) if isinstance(p, dict) else (float(p[0]), float(p[1]))
        except (KeyError, TypeError, ValueError, IndexError):
            raise PolygonError("Each zone point needs an x and a y.") from None
        if not (0.0 <= x <= 1.0 and 0.0 <= y <= 1.0):
            raise PolygonError("Zone points must lie inside the camera picture.")
        pts.append((x, y))
    if abs(area(pts)) < MIN_AREA:
        raise PolygonError("That zone is too small. Draw it around the whole area.")
    return pts


def area(poly: list[Point]) -> float:
    s = 0.0
    for i, (x1, y1) in enumerate(poly):
        x2, y2 = poly[(i + 1) % len(poly)]
        s += x1 * y2 - x2 * y1
    return s / 2.0


def contains(poly: list[Point], pt: Point) -> bool:
    """Ray casting; points on the edge count as inside."""
    x, y = pt
    inside = False
    n = len(poly)
    j = n - 1
    for i in range(n):
        xi, yi = poly[i]
        xj, yj = poly[j]
        if (yi > y) != (yj > y):
            x_cross = (xj - xi) * (y - yi) / (yj - yi) + xi
            if x <= x_cross:
                inside = not inside
        j = i
    return inside


@dataclass(frozen=True)
class BBox:
    """Normalised box, x1<x2, y1<y2."""

    x1: float
    y1: float
    x2: float
    y2: float

    @property
    def w(self) -> float:
        return self.x2 - self.x1

    @property
    def h(self) -> float:
        return self.y2 - self.y1

    def foot_point(self, frac: float = 0.9) -> Point:
        """Phase 2 §12: where the person stands, not where their head is."""
        return ((self.x1 + self.x2) / 2.0, self.y1 + self.h * frac)

    def upper_body(self) -> "BBox":
        """Top ~55 % of the body, widened — where arms reach a shelf from."""
        pad = self.w * 0.25
        return BBox(max(0.0, self.x1 - pad), self.y1, min(1.0, self.x2 + pad), self.y1 + self.h * 0.55)


def bbox_polygon_overlap(box: BBox, poly: list[Point], samples: int = 6) -> float:
    """Fraction of a box's sample grid that falls inside a polygon (cheap overlap)."""
    hit = 0
    total = samples * samples
    for i in range(samples):
        for j in range(samples):
            px = box.x1 + box.w * (i + 0.5) / samples
            py = box.y1 + box.h * (j + 0.5) / samples
            if contains(poly, (px, py)):
                hit += 1
    return hit / total


def polygon_bounds(poly: list[Point]) -> BBox:
    xs = [p[0] for p in poly]
    ys = [p[1] for p in poly]
    return BBox(min(xs), min(ys), max(xs), max(ys))

File: android/app/src/main/java/com/boombiz/guard/ai/Geometry.kt
package com.boombiz.guard.ai

/**
 * Zone geometry on NORMALISED coordinates (0–1), same as the PC agent's
 * zones/geometry.py, so a zone drawn on the sub-stream fits any resolution.
 */
data class Pt(val x: Float, val y: Float)

data class BBox(val x1: Float, val y1: Float, val x2: Float, val y2: Float) {
    val w get() = x2 - x1
    val h get() = y2 - y1

    /** Where the person stands, not where their head is. */
    fun foot(frac: Float = 0.9f) = Pt((x1 + x2) / 2f, y1 + h * frac)

    /** Top ~55 % of the body, widened — where arms reach a shelf from. */
    fun upperBody(): BBox {
        val pad = w * 0.25f
        return BBox(maxOf(0f, x1 - pad), y1, minOf(1f, x2 + pad), y1 + h * 0.55f)
    }
}

/** Share of a box's 6×6 sample grid inside a polygon (cheap overlap, PC geometry.py). */
fun overlap(box: BBox, poly: List<Pt>, samples: Int = 6): Float {
    var hit = 0
    for (i in 0 until samples) for (j in 0 until samples) {
        if (contains(poly, Pt(box.x1 + box.w * (i + 0.5f) / samples, box.y1 + box.h * (j + 0.5f) / samples))) hit++
    }
    return hit.toFloat() / (samples * samples)
}

/** How much of a person's arm reach overlaps a shelf. */
fun reachOverlap(box: BBox, shelf: List<Pt>) = overlap(box.upperBody(), shelf)

fun iou(a: BBox, b: BBox): Float {
    val ix = maxOf(0f, minOf(a.x2, b.x2) - maxOf(a.x1, b.x1))
    val iy = maxOf(0f, minOf(a.y2, b.y2) - maxOf(a.y1, b.y1))
    val inter = ix * iy
    val union = a.w * a.h + b.w * b.h - inter
    return if (union > 0f) inter / union else 0f
}

/** Ray casting; points on the edge count as inside. */
fun contains(poly: List<Pt>, p: Pt): Boolean {
    var inside = false
    var j = poly.size - 1
    for (i in poly.indices) {
        val (xi, yi) = poly[i]
        val (xj, yj) = poly[j]
        if ((yi > p.y) != (yj > p.y)) {
            val xCross = (xj - xi) * (p.y - yi) / (yj - yi) + xi
            if (p.x <= xCross) inside = !inside
        }
        j = i
    }
    return inside
}

fun area(poly: List<Pt>): Float {
    var s = 0f
    for (i in poly.indices) {
        val a = poly[i]
        val b = poly[(i + 1) % poly.size]
        s += a.x * b.y - b.x * a.y
    }
    return s / 2f
}

/**
 *   RESTRICTED  anyone stepping in (feet inside) alerts, day or night
 *   IGNORE      people standing here are never counted
 *   SHELF       products: a hand reaching in and the shelf changing is watched
 *   EXIT        the shop door: leaving with an unchecked shelf change alerts
 *   CASHIER     the pay point: passing it lowers the confidence of an exit alert
 */
enum class ZoneType(val label: String, val defaultName: String) {
    RESTRICTED("restricted", "Restricted area"),
    IGNORE("ignore", "Ignored area"),
    SHELF("shelf", "Shelf"),
    EXIT("exit", "Exit"),
    CASHIER("cashier", "Cashier"),
}

data class Zone(val id: Long, val cameraId: Long, val name: String, val type: ZoneType, val polygon: List<Pt>) {
    fun encode(): String = polygon.joinToString(";") { "${it.x},${it.y}" }

    companion object {
        const val MIN_AREA = 0.0005f

        fun decode(s: String): List<Pt> = s.split(";").filter { it.isNotBlank() }.map {
            val (x, y) = it.split(",")
            Pt(x.toFloat(), y.toFloat())
        }

        /** A dragged rectangle → polygon, clamped to the picture. Null when it's a mis-tap. */
        fun rect(a: Pt, b: Pt): List<Pt>? {
            val x1 = minOf(a.x, b.x).coerceIn(0f, 1f); val x2 = maxOf(a.x, b.x).coerceIn(0f, 1f)
            val y1 = minOf(a.y, b.y).coerceIn(0f, 1f); val y2 = maxOf(a.y, b.y).coerceIn(0f, 1f)
            val poly = listOf(Pt(x1, y1), Pt(x2, y1), Pt(x2, y2), Pt(x1, y2))
            return if (kotlin.math.abs(area(poly)) < MIN_AREA) null else poly
        }
    }
}
